src/software/ai/navigator/path_planner/global_path_planner.cpp

```diff
@@ -0,0 +1,59 @@
+#include "software/ai/navigator/path_planner/global_path_planner.h"
+
+GlobalPathPlannerFactory::GlobalPathPlannerFactory(
+    const std::shared_ptr<const RobotNavigationObstacleConfig> navigation_obstacle_config,
+    const World &world, const Rectangle &navigable_area)
+{
+    RobotNavigationObstacleFactory obstacle_factory =
+        RobotNavigationObstacleFactory(navigation_obstacle_config);
+    std::vector<MotionConstraint> all_constraints = allValuesMotionConstraint();
+
+    // The idea of this is similar to Gray codes
+    // (https://en.wikipedia.org/wiki/Gray_code#History_and_practical_application). The
+    // idea is that each bit in the counter represents a MotionConstraint and whether it
+    // is on and off. By cycling through every combination of bits, we'll consequently
+    // also have every combination of MotionConstraints
+    // (https://www.geeksforgeeks.org/generate-n-bit-gray-codes/)
+    for (unsigned counter = 0; counter < std::pow(2, all_constraints.size()); ++counter)
+    {
+        std::set<MotionConstraint> motion_constraint_obstacles;
+
+        // Use the value of the counter and bit arithmetic to get the motion constraint
+        // obstacles out
+        unsigned constraint_bits = counter;
+        for (unsigned j = 0; j < all_constraints.size(); ++j, constraint_bits >>= 1)
+        {
+            if (constraint_bits & 1)
+            {
+                motion_constraint_obstacles.emplace(all_constraints[j]);
+            }
+        }
+
+        auto obstacles = obstacle_factory.createFromMotionConstraints(
+            motion_constraint_obstacles, world);
```

Comment:
This won't work for `AVOID_BALL_PLACEMENT_INTERFERENCE` because the ball placement point changes and its constructed from the position of the ball

I think this potentially shouldn't be a motion constraint, and rather a soft rule that the enemy ball placement play follows

We can obv defer that work by opening a ticket, thoughts @itsarune @nimazareian @jonathanlew ?

Reply:
That's a good point, all motion constraints other than `HALF_METER_AROUND_BALL` and `AVOID_BALL_PLACEMENT_INTERFERENCE` seem to be static (can be handled by global path planner). HRVO can support  `HALF_METER_AROUND_BALL` pretty easily if we include it in the primitive, and with added support for other types of obstacles it should be able to support `AVOID_BALL_PLACEMENT_INTERFERENCE`. 
But I agree that this may be out of scope for this PR or the current HRVO integration PR.

Reply:
opening a ticket is fine, enemy ball placement won't work tho, so we'll need to fix it asap

Reply:
a thought is that we can have static motion constraints and dynamic motion constraints

Reply:
yeah makes sense, then we pass those motion constraints down through the move primitive

Reply:
yes, those motion constraints won't work, I'll just link an issue to it and we'll figure it out later